lib/Dialect/FIRRTL/InnerSymbolTable.cpp

```diff
@@ -30,38 +30,100 @@ InnerSymbolTable::InnerSymbolTable(Operation *op) {
+    // TODO: Add fields per port, once they work that way (use addSyms)
+    if (auto mod = dyn_cast<FModuleLike>(curOp)) {
+      for (const auto &p : llvm::enumerate(mod.getPorts()))
+        if (auto sym = p.value().sym; sym && !sym.getValue().empty())
+          addSym(p.value().sym, InnerSymTarget(p.index(), curOp));
+    }
   });
 }
 
-/// Look up a symbol with the specified name, returning null if no such name
-/// exists. Names never include the @ on them.
-Operation *InnerSymbolTable::lookup(StringRef name) const {
+/// Look up a symbol with the specified name, returning empty InnerSymTarget if
+/// no such name exists. Names never include the @ on them.
+InnerSymTarget InnerSymbolTable::lookup(StringRef name) const {
   return lookup(StringAttr::get(innerSymTblOp->getContext(), name));
 }
-Operation *InnerSymbolTable::lookup(StringAttr name) const {
+InnerSymTarget InnerSymbolTable::lookup(StringAttr name) const {
   return symbolTable.lookup(name);
 }
 
+/// Look up a symbol with the specified name, returning null if no such
+/// name exists or doesn't target just an operation.
+Operation *InnerSymbolTable::lookupOp(StringRef name) const {
+  return lookupOp(StringAttr::get(innerSymTblOp->getContext(), name));
+}
+Operation *InnerSymbolTable::lookupOp(StringAttr name) const {
+  auto result = lookup(name);
+  if (result.isOpOnly())
+    return result.getOp();
+  return nullptr;
+}
+
 /// Get InnerSymbol for an operation.
 StringAttr InnerSymbolTable::getInnerSymbol(Operation *op) {
   if (auto innerSymOp = dyn_cast<InnerSymbolOpInterface>(op))
     return innerSymOp.getInnerNameAttr();
   return {};
 }
 
-/// Return an InnerRef to the given operation.
-hw::InnerRefAttr InnerSymbolTable::getInnerRef(Operation *op) {
-  assert(op->getParentWithTrait<OpTrait::InnerSymbolTable>() == innerSymTblOp);
-  return hw::InnerRefAttr::get(SymbolTable::getSymbolName(innerSymTblOp),
-                               getInnerSymbol(op));
+/// Get InnerSymbol for a target.  Be robust to queries on unexpected
+/// operations to avoid users needing to know the details.
+StringAttr InnerSymbolTable::getInnerSymbol(InnerSymTarget target) {
+  // Assert on misuse, but try to handle queries otherwise.
+  assert(target);
+
+  InnerSymAttr base;
+  if (target.isPort()) {
+    auto mod = dyn_cast<FModuleLike>(target.getOp());
+    if (!mod)
+      return {};
+    assert(target.getPort() < mod.getNumPorts());
+    // TODO: update this when ports support per-field symbols
+    // base = mod.getPortSymbolAttr(target.getPort());
+    auto sym = mod.getPortSymbolAttr(target.getPort());
+    // Workaround quirk with empty string for no symbol on ports.
+    if (sym && sym.getValue().empty())
+      return {};
+    return sym;
+  } else {
+    // InnerSymbols only supported if op implements the interface.
+    auto symOp = dyn_cast<InnerSymbolOpInterface>(target.getOp());
+    if (!symOp)
+      return {};
+    base = symOp.getInnerSymAttr();
+  }
```

Comment:
Nit: in-line the `else` given that the `if` always returns.

Reply:
You're right.  This was written this way (with `base=`) to anticipate per-field port symbol support (which would also set an appropriate `base`, and not return (the if/else being responsible for initializing `base` or exiting on invalid), hence the `else`).

But if this is to be merged, let's make it look nice :+1:.  Not tricky enough to warrant that sort of placeholder code.